File: TypeScript/nestshop/src/friend/friend.service.ts
import { HttpStatus, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { In, Repository } from 'typeorm';
import { PaginationResponseDto } from '../common/dto/pagination.dto';
import { BusinessException } from '../common/exceptions/business.exception';
import { User } from '../user/entities/user.entity';
import { FriendPaginationQueryDto } from './dto/friend-pagination-query.dto';
import { FriendActivity } from './entities/friend-activity.entity';
import { FriendBlock } from './entities/friend-block.entity';
import { Friendship, FriendshipStatus } from './entities/friendship.entity';

@Injectable()
export class FriendService {
  constructor(
    @InjectRepository(Friendship)
    private friendshipRepository: Repository<Friendship>,
    @InjectRepository(FriendBlock)
    private friendBlockRepository: Repository<FriendBlock>,
    @InjectRepository(FriendActivity)
    private friendActivityRepository: Repository<FriendActivity>,
    @InjectRepository(User)
    private userRepository: Repository<User>,
  ) {}

  async requestFriend(userId: number, targetUserId: number) {
    await this.ensureUsers(userId, targetUserId);
    this.ensureNotSelf(userId, targetUserId);
    await this.ensureNotBlocked(userId, targetUserId);

    const existing = await this.findFriendshipPair(userId, targetUserId);
    if (existing && existing.status === FriendshipStatus.ACCEPTED) {
      throw new BusinessException('VALIDATION_FAILED', HttpStatus.CONFLICT, '이미 친구입니다.');
    }

    if (existing && existing.status === FriendshipStatus.PENDING) {
      throw new BusinessException('VALIDATION_FAILED', HttpStatus.CONFLICT, '이미 친구 요청이 존재합니다.');
    }

    if (existing) {
      existing.requesterId = userId;
      existing.addresseeId = targetUserId;
      existing.status = FriendshipStatus.PENDING;
      existing.respondedAt = null;
      await this.friendshipRepository.save(existing);
    } else {
      const friendship = this.friendshipRepository.create({
        requesterId: userId,
        addresseeId: targetUserId,
        status: FriendshipStatus.PENDING,
        respondedAt: null,
      });
      await this.friendshipRepository.save(friendship);
    }

    return { success: true, message: '친구 요청을 보냈습니다.' };
  }

  async acceptRequest(userId: number, friendshipId: number) {
    const friendship = await this.friendshipRepository.findOne({ where: { id: friendshipId } });
    if (!friendship) {
      throw new BusinessException('RESOURCE_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    if (friendship.addresseeId !== userId) {
      throw new BusinessException('AUTH_FORBIDDEN', HttpStatus.FORBIDDEN);
    }

    if (friendship.status !== FriendshipStatus.PENDING) {
      throw new BusinessException('VALIDATION_FAILED', HttpStatus.BAD_REQUEST, '대기중인 요청만 수락할 수 있습니다.');
    }

    friendship.status = FriendshipStatus.ACCEPTED;
    friendship.respondedAt = new Date();
    await this.friendshipRepository.save(friendship);

    await this.createActivity(friendship.requesterId, 'FRIEND_ACCEPTED', `유저 ${userId}님과 친구가 되었습니다.`, {
      friendUserId: userId,
    });
    await this.createActivity(userId, 'FRIEND_ACCEPTED', `유저 ${friendship.requesterId}님과 친구가 되었습니다.`, {
      friendUserId: friendship.requesterId,
    });

    return { success: true, message: '친구 요청을 수락했습니다.' };
  }

  async rejectRequest(userId: number, friendshipId: number) {
    const friendship = await this.friendshipRepository.findOne({ where: { id: friendshipId } });
    if (!friendship) {
      throw new BusinessException('RESOURCE_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    if (friendship.addresseeId !== userId) {
      throw new BusinessException('AUTH_FORBIDDEN', HttpStatus.FORBIDDEN);
    }

    if (friendship.status !== FriendshipStatus.PENDING) {
      throw new BusinessException('VALIDATION_FAILED', HttpStatus.BAD_REQUEST, '대기중인 요청만 거절할 수 있습니다.');
    }

    friendship.status = FriendshipStatus.REJECTED;
    friendship.respondedAt = new Date();
    await this.friendshipRepository.save(friendship);

    return { success: true, message: '친구 요청을 거절했습니다.' };
  }

  async getFriends(userId: number, query: FriendPaginationQueryDto) {
    const [rows, total] = await this.friendshipRepository.findAndCount({
      where: [
        { requesterId: userId, status: FriendshipStatus.ACCEPTED },
        { addresseeId: userId, status: FriendshipStatus.ACCEPTED },
      ],
      order: { updatedAt: 'DESC' },
      skip: query.skip,
      take: query.limit,
    });

    const friendIds = rows.map((row) => (row.requesterId === userId ? row.addresseeId : row.requesterId));
    const users = friendIds.length ? await this.userRepository.find({ where: { id: In(friendIds) } }) : [];
    const userMap = new Map(users.map((item) => [item.id, item]));

    const items = rows.map((row) => {
      const friendId = row.requesterId === userId ? row.addresseeId : row.requesterId;
      const friend = userMap.get(friendId);

      return {
        friendshipId: row.id,
        userId: friendId,
        nickname: friend?.nickname ?? null,
        profileImageUrl: friend?.profileImageUrl ?? null,
        status: row.status,
        since: row.respondedAt,
      };
    });

    return new PaginationResponseDto(items, total, query.page, query.limit);
  }

  async getReceivedRequests(userId: number, query: FriendPaginationQueryDto) {
    return this.getRequestList('received', userId, query);
  }

  async getSentRequests(userId: number, query: FriendPaginationQueryDto) {
    return this.getRequestList('sent', userId, query);
  }

  async getFeed(userId: number, query: FriendPaginationQueryDto) {
    const friends = await this.friendshipRepository.find({
      where: [
        { requesterId: userId, status: FriendshipStatus.ACCEPTED },
        { addresseeId: userId, status: FriendshipStatus.ACCEPTED },
      ],
    });

    const friendIds = friends.map((row) => (row.requesterId === userId ? row.addresseeId : row.requesterId));

    if (!friendIds.length) {
      return new PaginationResponseDto([], 0, query.page, query.limit);
    }

    const [activities, total] = await this.friendActivityRepository.findAndCount({
      where: { userId: In(friendIds) },
      order: { createdAt: 'DESC' },
      skip: query.skip,
      take: query.limit,
    });

    const users = await this.userRepository.find({ where: { id: In(friendIds) } });
    const userMap = new Map(users.map((item) => [item.id, item]));

    const items = activities.map((item) => ({
      id: item.id,
      userId: item.userId,
      nickname: userMap.get(item.userId)?.nickname ?? null,
      profileImageUrl: userMap.get(item.userId)?.profileImageUrl ?? null,
      type: item.type,
      message: item.message,
      metadata: item.metadata,
      createdAt: item.createdAt,
    }));

    return new PaginationResponseDto(items, total, query.page, query.limit);
  }

  async blockUser(userId: number, targetUserId: number) {
    await this.ensureUsers(userId, targetUserId);
    this.ensureNotSelf(userId, targetUserId);

    const existing = await this.friendBlockRepository.findOne({ where: { userId, blockedUserId: targetUserId } });
    if (existing) {
      return { success: true, message: '이미 차단한 사용자입니다.' };
    }

    const block = this.friendBlockRepository.create({ userId, blockedUserId: targetUserId });
    await this.friendBlockRepository.save(block);

    const friendship = await this.findFriendshipPair(userId, targetUserId);
    if (friendship) {
      await this.friendshipRepository.softDelete({ id: friendship.id });
    }

    return { success: true, message: '사용자를 차단했습니다.' };
  }

  async unblockUser(userId: number, targetUserId: number) {
    const block = await this.friendBlockRepository.findOne({ where: { userId, blockedUserId: targetUserId } });
    if (!block) {
      throw new BusinessException('RESOURCE_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    await this.friendBlockRepository.softDelete({ id: block.id });
    return { success: true, message: '차단을 해제했습니다.' };
  }

  async removeFriend(userId: number, targetUserId: number) {
    const friendship = await this.findFriendshipPair(userId, targetUserId);
    if (!friendship || friendship.status !== FriendshipStatus.ACCEPTED) {
      throw new BusinessException('RESOURCE_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    await this.friendshipRepository.softDelete({ id: friendship.id });
    return { success: true, message: '친구를 삭제했습니다.' };
  }

  private async getRequestList(
    type: 'received' | 'sent',
    userId: number,
    query: FriendPaginationQueryDto,
  ) {
    const where =
      type === 'received'
        ? { addresseeId: userId, status: FriendshipStatus.PENDING }
        : { requesterId: userId, status: FriendshipStatus.PENDING };

    const [rows, total] = await this.friendshipRepository.findAndCount({
      where,
      order: { createdAt: 'DESC' },
      skip: query.skip,
      take: query.limit,
    });

    const targetIds = rows.map((row) => (type === 'received' ? row.requesterId : row.addresseeId));
    const users = targetIds.length ? await this.userRepository.find({ where: { id: In(targetIds) } }) : [];
    const userMap = new Map(users.map((item) => [item.id, item]));

    const items = rows.map((row) => {
      const targetId = type === 'received' ? row.requesterId : row.addresseeId;
      const target = userMap.get(targetId);
      return {
        friendshipId: row.id,
        userId: targetId,
        nickname: target?.nickname ?? null,
        profileImageUrl: target?.profileImageUrl ?? null,
        requestedAt: row.createdAt,
      };
    });

    return new PaginationResponseDto(items, total, query.page, query.limit);
  }

  private async findFriendshipPair(userId: number, targetUserId: number) {
    return this.friendshipRepository.findOne({
      where: [
        { requesterId: userId, addresseeId: targetUserId },
        { requesterId: targetUserId, addresseeId: userId },
      ],
    });
  }

  private async ensureUsers(userId: number, targetUserId: number) {
    const users = await this.userRepository.find({ where: { id: In([userId, targetUserId]) } });
    if (users.length !== 2) {
      throw new BusinessException('USER_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
  }

  private ensureNotSelf(userId: number, targetUserId: number) {
    if (userId === targetUserId) {
      throw new BusinessException('VALIDATION_FAILED', HttpStatus.BAD_REQUEST, '자기 자신에게는 수행할 수 없습니다.');
    }
  }

  private async ensureNotBlocked(userId: number, targetUserId: number) {
    const block = await this.friendBlockRepository.findOne({
      where: [
        { userId, blockedUserId: targetUserId },
        { userId: targetUserId, blockedUserId: userId },
      ],
    });

    if (block) {
      throw new BusinessException('VALIDATION_FAILED', HttpStatus.BAD_REQUEST, '차단된 사용자와는 요청할 수 없습니다.');
    }
  }

  private async createActivity(userId: number, type: string, message: string, metadata: Record<string, unknown>) {
    const activity = this.friendActivityRepository.create({
      userId,
      type,
      message,
      metadata,
    });

    await this.friendActivityRepository.save(activity);
  }
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/friend/FriendService.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.friend

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.PbShopException
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.StubResponse
import io.ktor.http.HttpStatusCode

class FriendService(
    private val repository: FriendRepository,
) {
    fun requestFriend(userId: Int, targetUserId: Int): StubResponse {
        ensureUser(userId)
        ensureUser(targetUserId)
        ensureNotSelf(userId, targetUserId)
        ensureNotBlocked(userId, targetUserId)
        val existing = repository.findFriendshipPair(userId, targetUserId)
        if (existing?.status == "ACCEPTED") {
            throw PbShopException(HttpStatusCode.Conflict, "VALIDATION_ERROR", "이미 친구입니다.")
        }
        if (existing?.status == "PENDING") {
            throw PbShopException(HttpStatusCode.Conflict, "VALIDATION_ERROR", "이미 친구 요청이 존재합니다.")
        }
        repository.saveFriendship(userId, targetUserId, "PENDING", existing?.id)
        return StubResponse(data = mapOf("message" to "친구 요청을 보냈습니다."))
    }

    fun acceptRequest(userId: Int, friendshipId: Int): StubResponse {
        val friendship = requireFriendship(friendshipId)
        if (friendship.addresseeId != userId) {
            throw PbShopException(HttpStatusCode.Forbidden, "FORBIDDEN", "수락 권한이 없습니다.")
        }
        if (friendship.status != "PENDING") {
            throw PbShopException(HttpStatusCode.BadRequest, "INVALID_STATUS_TRANSITION", "대기중인 요청만 수락할 수 있습니다.")
        }
        repository.saveFriendship(friendship.requesterId, friendship.addresseeId, "ACCEPTED", friendship.id)
        repository.createActivity(friendship.requesterId, "FRIEND_ACCEPTED", "유저 ${friendship.addresseeId}님과 친구가 되었습니다.", mapOf("friendUserId" to friendship.addresseeId.toString()))
        repository.createActivity(friendship.addresseeId, "FRIEND_ACCEPTED", "유저 ${friendship.requesterId}님과 친구가 되었습니다.", mapOf("friendUserId" to friendship.requesterId.toString()))
        return StubResponse(data = mapOf("message" to "친구 요청을 수락했습니다."))
    }

    fun rejectRequest(userId: Int, friendshipId: Int): StubResponse {
        val friendship = requireFriendship(friendshipId)
        if (friendship.addresseeId != userId) {
            throw PbShopException(HttpStatusCode.Forbidden, "FORBIDDEN", "거절 권한이 없습니다.")
        }
        if (friendship.status != "PENDING") {
            throw PbShopException(HttpStatusCode.BadRequest, "INVALID_STATUS_TRANSITION", "대기중인 요청만 거절할 수 있습니다.")
        }
        repository.saveFriendship(friendship.requesterId, friendship.addresseeId, "REJECTED", friendship.id)
        return StubResponse(data = mapOf("message" to "친구 요청을 거절했습니다."))
    }

    fun listFriends(userId: Int, page: Int, limit: Int): StubResponse {
        val result = repository.listFriends(userId, normalizePage(page), normalizeLimit(limit))
        return StubResponse(
            data = result.items.map { friendshipPayload(it, userId) },
            meta = pageMeta(page, limit, result.totalCount),
        )
    }

    fun receivedRequests(userId: Int, page: Int, limit: Int): StubResponse {
        val normalizedPage = normalizePage(page)
        val normalizedLimit = normalizeLimit(limit)
        val result = repository.listReceivedRequests(userId, normalizedPage, normalizedLimit)
        return StubResponse(data = result.items.map(::requestPayload), meta = pageMeta(normalizedPage, normalizedLimit, result.totalCount))
    }

    fun sentRequests(userId: Int, page: Int, limit: Int): StubResponse {
        val normalizedPage = normalizePage(page)
        val normalizedLimit = normalizeLimit(limit)
        val result = repository.listSentRequests(userId, normalizedPage, normalizedLimit)
        return StubResponse(data = result.items.map(::requestPayload), meta = pageMeta(normalizedPage, normalizedLimit, result.totalCount))
    }

    fun feed(userId: Int, page: Int, limit: Int): StubResponse {
        val normalizedPage = normalizePage(page)
        val normalizedLimit = normalizeLimit(limit)
        val result = repository.listFeed(userId, normalizedPage, normalizedLimit)
        return StubResponse(data = result.items.map(::activityPayload), meta = pageMeta(normalizedPage, normalizedLimit, result.totalCount))
    }

    fun blockUser(userId: Int, targetUserId: Int): StubResponse {
        ensureUser(userId)
        ensureUser(targetUserId)
        ensureNotSelf(userId, targetUserId)
        repository.createBlock(userId, targetUserId)
        repository.deleteFriendshipPair(userId, targetUserId)
        return StubResponse(data = mapOf("message" to "사용자를 차단했습니다."))
    }

    fun unblockUser(userId: Int, targetUserId: Int): StubResponse {
        if (!repository.deleteBlock(userId, targetUserId)) {
            throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "차단 정보를 찾을 수 없습니다.")
        }
        return StubResponse(data = mapOf("message" to "차단을 해제했습니다."))
    }

    fun removeFriend(userId: Int, targetUserId: Int): StubResponse {
        val friendship = repository.findFriendshipPair(userId, targetUserId)
        if (friendship == null || friendship.status != "ACCEPTED") {
            throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "친구 관계를 찾을 수 없습니다.")
        }
        repository.deleteFriendshipPair(userId, targetUserId)
        return StubResponse(data = mapOf("message" to "친구를 삭제했습니다."))
    }

    private fun ensureUser(userId: Int) {
        if (!repository.userExists(userId)) {
            throw PbShopException(HttpStatusCode.NotFound, "USER_NOT_FOUND", "사용자를 찾을 수 없습니다.")
        }
    }

    private fun ensureNotSelf(userId: Int, targetUserId: Int) {
        if (userId == targetUserId) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "자기 자신에게는 수행할 수 없습니다.")
        }
    }

    private fun ensureNotBlocked(userId: Int, targetUserId: Int) {
        if (repository.isBlockedBetween(userId, targetUserId)) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "차단된 사용자와는 요청할 수 없습니다.")
        }
    }

    private fun requireFriendship(friendshipId: Int): FriendshipRecord =
        repository.findFriendshipById(friendshipId)
            ?: throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "친구 요청을 찾을 수 없습니다.")

    private fun friendshipPayload(record: FriendshipRecord, currentUserId: Int): Map<String, Any?> {
        val friend = if (record.requesterId == currentUserId) record.addressee else record.requester
        return mapOf(
            "id" to record.id,
            "status" to record.status,
            "createdAt" to record.createdAt.toString(),
            "updatedAt" to record.updatedAt.toString(),
            "friend" to userPayload(friend),
        )
    }

    private fun requestPayload(record: FriendshipRecord): Map<String, Any?> =
        mapOf(
            "id" to record.id,
            "status" to record.status,
            "createdAt" to record.createdAt.toString(),
            "requester" to userPayload(record.requester),
            "addressee" to userPayload(record.addressee),
        )

    private fun activityPayload(record: FriendActivityRecord): Map<String, Any?> =
        mapOf(
            "id" to record.id,
            "userId" to record.userId,
            "type" to record.type,
            "message" to record.message,
            "metadata" to record.metadata,
            "createdAt" to record.createdAt.toString(),
            "user" to userPayload(record.user),
        )

    private fun userPayload(record: FriendUserRecord): Map<String, Any?> =
        mapOf(
            "id" to record.id,
            "email" to record.email,
            "name" to record.name,
            "nickname" to record.nickname,
            "profileImageUrl" to record.profileImageUrl,
            "role" to record.role,
        )

    private fun normalizePage(page: Int): Int = if (page > 0) page else 1

    private fun normalizeLimit(limit: Int): Int = limit.coerceIn(1, 100).takeIf { it > 0 } ?: 20

    private fun pageMeta(page: Int, limit: Int, totalCount: Int): Map<String, Int> =
        mapOf(
            "page" to page,
            "limit" to limit,
            "totalCount" to totalCount,
            "totalPages" to if (totalCount == 0) 0 else ((totalCount - 1) / limit) + 1,
        )
}
